examples/erc1155/lib.rs

```diff
@@ -0,0 +1,687 @@
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_batch_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_ids: Vec<TokenId>,
+        values: Vec<Balance>,
+        data: Vec<u8>,
+    );
+}
+
+#[ink::contract]
+mod erc1155 {
+    use super::*;
+
+    use ink_env::call::{
+        build_call,
+        utils::ReturnType,
+        ExecutionInput,
+        Selector,
+    };
+    use ink_prelude::collections::BTreeMap;
+
+    /// Indicate that a token transfer has occured.
+    ///
+    /// This must be emitted even if a zero value transfer occurs.
+    #[ink(event)]
+    pub struct TransferSingle {
+        #[ink(topic)]
+        operator: AccountId,
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+        token_id: TokenId,
+        value: Balance,
+    }
+
+    /// Indicate that an approval event has happened.
+    #[ink(event)]
+    pub struct ApprovalForAll {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        operator: AccountId,
+        approved: bool,
+    }
+
+    /// Indicate that a token's URI has been updated.
+    #[ink(event)]
+    pub struct Uri {
+        value: ink_prelude::string::String,
+        #[ink(topic)]
+        token_id: TokenId,
+    }
+
+    /// An ERC-1155 contract.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Contract {
+        /// Tracks the balances of accounts across the different tokens that they might be holding.
+        balances: BTreeMap<(AccountId, TokenId), Balance>,
+        /// Which accounts (called operators) have been approved to spend funds on behalf of an owner.
+        ///
+        /// Note that the mapping is Set<(Owner, Operator)>.
+        approvals: BTreeMap<(AccountId, AccountId), ()>,
+        /// A unique identifier for the tokens which have been minted (and are therefore supported)
+        /// by this contract.
+        token_id_nonce: TokenId,
+    }
+
+    impl Contract {
+        /// Initialize a default instance of this ERC-1155 implementation.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Default::default()
+        }
+
+        /// Create the initial supply for a token.
+        ///
+        /// The initial supply will be provided to the caller (a.k.a the minter), and the
+        /// `token_id` will be assigned by the smart contract.
+        ///
+        /// Note that as implemented anyone can create tokens. If you were to deploy this to a
+        /// production environment you'd probably want to lock down the addresses that are allowed
+        /// to create tokens.
+        #[ink(message)]
+        pub fn create(&mut self, value: Balance, _data: Vec<u8>) -> TokenId {
+            // Given that TokenId is a `u128` the likelihood of this overflowing is pretty slim.
+            self.token_id_nonce += 1;
+            self.balances
+                .insert((self.env().caller(), self.token_id_nonce), value);
```

Comment:
maybe capture `self.env().caller()` into some local binding since you access it 3 times in this function alone.

Reply:
That's is only the case for `ink_storage` `collections` and `lazy` data structures.
Doing this for things like `self.env().caller()` and others would require a more elaborate process in order to produce optimal code and with our current process we couldn't produce optimal code so we do not even try and instead let the users do their job of not repeating themselves. However, in the future it might be possible to implement lints that help users with these things via `clippy` plugins.